src/System.Linq/src/System/Linq/OrderedEnumerable.cs

```diff
@@ -26,10 +27,21 @@ public IEnumerator<TElement> GetEnumerator()
             Buffer<TElement> buffer = new Buffer<TElement>(_source);
             if (buffer._count > 0)
             {
-                int[] map = SortedMap(buffer);
-                for (int i = 0; i < buffer._count; i++)
+                int[] map = null;
+                try
                 {
-                    yield return buffer._items[map[i]];
+                    map = SortedMap(buffer);
+                    for (int i = 0; i < buffer._count; i++)
+                    {
+                        yield return buffer._items[map[i]];
+                    }
+                }
+                finally
```

Comment:
do we strictly need to guarantee the return of the rented array no matter what? 
Can we just return without putting this into a finally?   

Using finally seems a bit heave, especially in an iterator. 
Also, depending on the pool implementation, it seems it might expose the pool to a potential corruption if someone Disposes the iterator twice concurrently on separate threads.